keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md

```diff
@@ -0,0 +1,296 @@
+    * Conversions from the external type to the internal type will be kept in `{internal_api_path}/{external_version}`, like for `k8s.io/api`
+        * Defaulting code will be kept in this package, next to the conversions
+* Create a "shared types"-package with structs generic to all or many componentconfig API groups, in the `k8s.io/apimachinery`, `k8s.io/apiserver` and `k8s.io/controller-manager` repos, depending on the struct. 
+    * Location: `k8s.io/{apimachinery,apiserver,controller-manager}/pkg/apis/config/{,v1alpha1}`
+    * These aren’t "real" API groups, but have both internal and external versions
+    * Defaulting, conversion and internal types are published to the staging repo.
+* Remove the monolithic `componentconfig/v1alpha1` API group
+* Enable the staging bot to create the Github repos
+
+### Migration strategy per component or k8s.io repo
+
+#### k8s.io/apimachinery changes
+
+* **Not a "real" API group, instead shared packages only with both external and internal types.**
+* **External Package with defaulting & conversions**: `k8s.io/apimachinery/pkg/apis/config/v1alpha1/types.go`
+* **Internal Package**: `k8s.io/apimachinery/pkg/apis/config/types.go`
+* Structs to be hosted:
+    * ClientConnectionConfiguration
+* Assignee: ?
+
+#### k8s.io/apiserver changes
+
+* **Not a "real" API group, instead shared packages only with both external and internal types.**
+* **External Package with defaulting & conversions**: `k8s.io/apiserver/pkg/apis/config/v1alpha1/types.go`
+* **Internal Package**: `k8s.io/apiserver/pkg/apis/config/types.go`
+* Structs to be hosted:
+    * LeaderElectionConfiguration
+    * DebuggingConfiguration
+    * later to be created: SecureServingConfiguration, AuthenticationConfiguration, AuthorizationConfiguration, etc.
+* Assignee: ?
+
+#### kubelet changes
+
+* **GroupVersionKind:** `kubelet.config.k8s.io/v1beta.KubeletConfiguration`
+* **External Package:** `k8s.io/kubelet/config/v1beta1/types.go`
+* **Internal Package:** `k8s.io/kubernetes/pkg/kubelet/apis/config/types.go`
+* **Internal Scheme:** `k8s.io/kubernetes/pkg/kubelet/apis/config/scheme/scheme.go`
+* **Conversions & defaulting Package:** `k8s.io/kubernetes/pkg/kubelet/apis/config/v1beta1/register.go`
+* **Future Internal Package:** `k8s.io/kubelet/pkg/apis/config/types.go`
+* Assignee: @mtaufen?
+
+#### kube-proxy changes
+
+* **GroupVersionKind**: `kubeproxy.config.k8s.io/v1alpha1.KubeProxyConfiguration`
+* **External Package**: `k8s.io/kube-proxy/config/v1alpha1/types.go`
+* **Internal Package**: `k8s.io/kubernetes/pkg/proxy/apis/config/types.go`
+* **Internal Scheme**: `k8s.io/kubernetes/pkg/proxy/apis/config/scheme/scheme.go`
+* **Conversions & defaulting Package:** `k8s.io/kubernetes/pkg/proxy/apis/config/v1alpha1/register.go`
+* **Future Internal Package:** `k8s.io/kube-proxy/pkg/apis/config/types.go`
+* Start referencing `ClientConnectionConfiguration` from the generic ComponentConfig packages
+* Assignee: ?
```

Comment:
@rramkumar1 or @dcbw or @danwinship  - anyone willing to be assigned this?  I don't know how much actual urgency is here, but it's about time to start getting the config cleaned up?

Reply:
@thockin You can assign it to here if all of other guys are busy :)

Reply:
Maybe I can join it ？

Reply:
Added @m1093782566. @hanxiaoshuai feel free to collaborate on this one of course.

Reply:
Thanks @sttts , I would collaborate with @hanxiaoshuai :)